hudi/src/main/scala/io/qbeast/spark/hudi/HudiRollupDataWriter.scala

```diff
@@ -0,0 +1,141 @@
+import org.apache.spark.sql.types.StructType
+import org.apache.spark.sql.DataFrame
+import org.apache.spark.sql.SparkSession
+
+/**
+ * Delta implementation of DataWriter that applies rollup to compact the files.
+ */
+object HudiRollupDataWriter extends RollupDataWriter {
+
+  override def write(
+      tableId: QTableID,
+      schema: StructType,
+      data: DataFrame,
+      tableChanges: TableChanges,
+      commitTime: String): IISeq[IndexFile] = {
+
+    if (data.isEmpty) return Seq.empty[IndexFile].toIndexedSeq
+
+    val statsTrackers = StatsTracker.getStatsTrackers
+
+    val extendedData = extendDataWithFileUUID(data, tableChanges)
+    val hudiData = extendDataWithHudiColumns(extendedData, commitTime)
+
+    val hudiSchema = schema
+      .add(StructField("_hoodie_commit_time", StringType, nullable = false))
+      .add(StructField("_hoodie_commit_seqno", StringType, nullable = false))
+      .add(StructField("_hoodie_record_key", StringType, nullable = false))
+      .add(StructField("_hoodie_partition_path", StringType, nullable = false))
+      .add(StructField("_hoodie_file_name", StringType, nullable = false))
+
+    val filesAndStats =
+      doWrite(tableId, hudiSchema, hudiData, tableChanges, statsTrackers)
+    val stats = filesAndStats.map(_._2)
+
+    processStats(stats, statsTrackers)
+
+    filesAndStats
+      .map(_._1)
+  }
+
+  private def processStats(
+      stats: IISeq[TaskStats],
+      statsTrackers: Seq[WriteJobStatsTracker]): Unit = {
+    val basicStatsBuilder = Seq.newBuilder[WriteTaskStats]
+    var endTime = 0L
+    stats.foreach(stats => {
+      basicStatsBuilder ++= stats.writeTaskStats.filter(_.isInstanceOf[BasicWriteTaskStats])
+      endTime = math.max(endTime, stats.endTime)
+    })
+    val basicStats = basicStatsBuilder.result()
+    statsTrackers.foreach(_.processStats(basicStats, endTime))
+  }
+
+  private def extendDataWithHudiColumns(
+      extendedData: DataFrame,
+      commitTime: String): DataFrame = {
+
+    def generateRecordKey(timestamp: String): UserDefinedFunction =
+      udf((groupdID: Int, rowId: Int) => {
+        s"${timestamp}_${groupdID}_$rowId"
+      })
+
+    def generateCommitSeqno(timestamp: String): UserDefinedFunction =
+      udf((groupdID: Int, rowId: Int) => {
+        s"${timestamp}_${groupdID}_$rowId"
+      })
+
+    def generateFilename(timestamp: String): UserDefinedFunction =
+      udf((uuid: String, groupdID: Int) => {
+        val token = s"$groupdID-${groupdID + 13}-0"
+        s"$uuid-0_${token}_$timestamp.parquet"
+      })
+
+    val distinctValues = extendedData
+      .select(col(QbeastColumns.fileUUIDColumnName))
+      .distinct()
+      .collect()
+      .map(_.getString(0))
+      .zipWithIndex
+      .toMap
+
+    val mappingBroadcast = SparkSession.active.sparkContext.broadcast(distinctValues)
+    val assignRowGroupUDF = udf((value: String) => mappingBroadcast.value.getOrElse(value, -1))
+
+    val windowSpecGroup = Window
```

Comment:
I still need to evaluate the impact of this. In principle, the window operation that uses partitionBy() is not as inefficient as the one without it:

Last week, I worked extensively on improving this part, as my initial implementation had two window operations: the one you see here and another one without partitionBy() (to calculate the GROUP_ID). Spark itself was showing warning messages for the second one.
```
WARN WindowExec: No Partition Defined for Window operation! Moving all data to a single partition, this can cause serious performance degradation.
```

I also spent a few days trying to figure out how Hudi generates this (GROUP_ID and ROW_ID) but couldn't find it. I got the impression that it was assigned arbitrarily.

I will pay attention to this during the upcoming tests and compare the times with and without executing this part

Reply:
After testing this part, I found it had excessive overhead, so I completely removed and redesigned it. In the latest commit, I introduced the new version (more similar to what Hudi does), which, after some experiments, shows no overhead.